pkg/pillar/cmd/zedagent/handlecertconfig.go

```diff
@@ -254,6 +295,16 @@ func getCertsFromController(ctx *zedagentContext) bool {
 		return false
 	}
 
+	// Did the certificate change?
+	_, err = os.Stat(types.ServerSigningCertFileName)
+	if err == nil {
+		oldCertBytes, err := ioutil.ReadFile(types.ServerSigningCertFileName)
+		if err == nil && bytes.Equal(oldCertBytes, certBytes) {
+			log.Functionf("getCertsFromController: unchanged cert")
+			return true // Succeeded
+		}
+	}
+
```

Comment:
Here, if Controller cert has expired, since we pass skipVerify=True above, we will end up in this line and see that the cert has not changed, and will return true here from getControllerCerts()? If that is the case, we will never retry again till the next day (or whatever the longer timer is)? Note that we will again trigger fetching cert upon seeing that the cert has expired while processing device config from the Controller, but since that is going to hit this line again, we are not going to overwrite it?  Am I missing something @eriknordmark ?

if err == nil && bytes.Equal(oldCertBytes, certBytes) && certIsValid() {
} 

or even 
skipping the check and overwriting it unconditionally below

might be a safer bet here IMHO

Reply:
The cert chain verification is done in that case by calling zedcloud.VerifyProtoSigningCertChain().

We pass SkipVerify=true since the cert payload is itself not signed using AuthContainer; that would introduce a chicken and egg problem if we did.

I didn't understand why we'd always want to write the file? That merely increases write traffic to the SSD, since we checked that the content is the same.

Reply:
Missed reading the call to VerifyProtoSigningCertChain() above. Pls ignore my comment